src/ice_nucleation.F90

```diff
@@ -0,0 +1,563 @@
+! Copyright (C) 2024 University of Illinois at Urbana-Champaign
+! Licensed under the GNU General Public License version 2 or (at your
+! option) any later version. See the file COPYING for details.
+
+!> \file
+! The pmc_ice_nucleation module.
+
+module pmc_ice_nucleation
+  use pmc_aero_state
+  use pmc_env_state
+  use pmc_aero_data
+  use pmc_util
+  use pmc_aero_particle
+  use pmc_constants
+  use pmc_rand
+
+  implicit none
+
+  !> Type code for an undefined or invalid immersion freezing scheme.
+  integer, parameter :: IMMERSION_FREEZING_SCHEME_INVALID = 0
+  !> Type code for constant ice nucleation rate (J_het) immersion freezing scheme.
+  integer, parameter :: IMMERSION_FREEZING_SCHEME_CONST = 1
+  !> Type code for the singular (INAS) immersion freezing scheme.
+  integer, parameter :: IMMERSION_FREEZING_SCHEME_SINGULAR = 2
+  !> Type code for the ABIFM immersion freezing scheme.
+  integer, parameter :: IMMERSION_FREEZING_SCHEME_ABIFM = 3
+
+  !> Whether to use the binned-tau leaping algorithm for time-dependent scheme.
+  logical :: do_speedup = .true.
+
+  interface ABIFM
+    module procedure ABIFM_particle
+    module procedure ABIFM_max
+  end interface
+
+contains
+
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+  !> Main subroutine for immersion freezing simulation.
+  subroutine immersion_freezing(aero_state, aero_data, env_state_initial, &
+    env_state_final, del_t, immersion_freezing_scheme_type, freezing_rate)
+    !do_coating, coating_spec, coating_ratio)
+    implicit none
+    !> Aerosol state.
+    type(aero_state_t), intent(inout) :: aero_state
+    !> Aerosol data.
+    type(aero_data_t), intent(in) :: aero_data
+    !> Environment state at the start of the timestep.
+    type(env_state_t), intent(in) :: env_state_initial
+    !> Environment state at the end of the timestep. The rel_humid
+    !> value will be ignored and overwritten with a new value.
+    type(env_state_t), intent(inout) :: env_state_final
+    !> Total time to integrate.
+    real(kind=dp), intent(in) :: del_t
+    !> Immersion freezing scheme type (e.g., 1:constant rate, 2:sigular, 3:ABIFM).
```

Comment:
Consider removing the e.g. portion as its specific to the parameters and not this subroutine. They could in theory change.